experiment/simulation/js/main.js

```diff
@@ -1 +1,559 @@
+  drawLine(left, 0, left, textHeight, 2, leftColor);
+  drawLine(right, 0, right, textHeight, 2, rightColor);
+  drawLine(0, down, width, down, 2, bottomColor);
+  drawLine(0, up, width, up, 2, topColor);
+  drawLine(0, textHeight, width, textHeight, 2, "black");
+  // mark the coordinates for the rectangular grid's intersection
+  makePoint(left, up, "red", "red");
+  makePoint(right, down, "red", "red");
+  makePoint(left, down, "red", "red");
+  makePoint(right, up, "red", "red");
+}
+function calculateSlope() {
+  if (point1[0] == point2[0]) {
+    if (point1[1] < point2[1]) {
+      return Number.MAX_SAFE_INTEGER;
+    } else {
+      return Number.MIN_SAFE_INTEGER;
+    }
+  }
+  return (point2[1] - point1[1]) / (point2[0] - point1[0]);
+}
+// function for clipping the points
+function clipPoint(point, edge) {
+  point[0] = parseFloat(point[0]);
+  point[1] = parseFloat(point[1]);
+  left = parseFloat(left);
+  right = parseFloat(right);
+  up = parseFloat(up);
+  down = parseFloat(down);
+  const pointCode = encodePoint(point[0], point[1]);
+  const slope = parseFloat(calculateSlope());
+  if (edge === "BOTTOM") {
+    if (pointCode & BOTTOM) {
+      if (slope !== 0) {
+        const x = point[0] + (1 / slope) * (down - point[1]);
+        const y = down;
+        return [x, y];
+      }
+    }
+  } else if (edge === "LEFT") {
+    if (pointCode & LEFT) {
+      if (
+        slope === Number.MAX_SAFE_INTEGER ||
+        slope === Number.MIN_SAFE_INTEGER
+      ) {
+        return [];
+      } else {
+        const x = left;
+        const y = point[1] + slope * (left - point[0]);
+        return [x, y];
+      }
+    }
+  } else if (edge === "RIGHT") {
+    if (pointCode & RIGHT) {
+      if (
+        slope === Number.MAX_SAFE_INTEGER ||
+        slope === Number.MIN_SAFE_INTEGER
+      ) {
+        return [];
+      } else {
+        const x = right;
+        const y = point[1] + slope * (right - point[0]);
+        return [x, y];
+      }
+    }
+  } else if (edge === "TOP") {
+    if (pointCode & TOP) {
+      if (slope !== 0) {
+        const x = point[0] + (1 / slope) * (up - point[1]);
+        const y = up;
+        return [x, y];
+      }
+    }
+  }
+  return [];
+}
+function pointBeingClipped() {
+  // whether point 1 is being clipped or the point 2
+  if (times_next_called >= 1 && times_next_called <= 8) {
+    return "point1";
+  } else if (times_next_called >= 9 && times_next_called <= 16) {
+    return "point2";
+  } else {
+    return null;
+  }
+}
+function clippingEdge() {
+  // returns the current edge against which the point is being clipped
+  const option = times_next_called % 8;
+  if (option === 1 || option === 2) {
+    return "LEFT";
+  } else if (option === 3 || option === 4) {
+    return "BOTTOM";
+  } else if (option === 5 || option === 6) {
+    return "RIGHT";
+  } else if (option === 7 || option === 0) {
+    return "TOP";
+  }
+}
+function highlightEdge(EDGE, width, color) {
```

Comment:
EDGE is a regular function argument, please name it using camelCase and not using all capitals